### ISSUE TITLE
Boot scripts only run after opening Termux app

### ISSUE THREAD

**Issue by ivanfor** [termux/termux-boot#47]
**Problem description**
Scripts in ~/.termux/boot seem to be run only after termux app is opened instead automatically after boot.
Is there anything i can do to troubleshoot?

**Steps to reproduce**

1. Install termux app (from f-droid in my case)
2. Start termux boot app (see the instructions)
3. Create ~/.termux/boot dir
4. Create the following script:
```
$ cat 00_autostart 
#!/data/data/com.termux/files/usr/bin/bash
termux-wake-lock
date > $HOME/date

```

**Expected behavior**
Create the $HOME/date file after boot

**Additional information**

* Termux application version: 0.95. Termux boot is version 0.7
* Android OS version: 9.0
* Device model: unbranded chinese tablet.

**Comment by Cjkeenan** [comment]
I am having this issue as well. I am trying to autostart an instance of node red. It seems to be acquiring the wake-lock properly, but then says 0 sessions. I would love to see a resolution to this.

**Comment by Grimler91** [comment]
@Cjkeenan for node-red you need to fix the shebang of the node-red script. 

It will not start an actual terminal session though, termux-boot expects the command to finish (or continue in the background)

**Comment by Cjkeenan** [comment]
interesting, is there a guide somewhere on how to fix this? I was just following the guide on Node-RED's own documentation for how to install on Android and they recommended autostarting via termux:boot

**Comment by Grimler91** [comment]
@Cjkeenan after installing node-red you see that the first line (the shebang) of the node-red script is `#!/usr/bin/env node` (check with `head $(which node-red)`. `/usr/bin/env` does not exist on android though, so the shebang should be changed to `#!/data/data/com.termux/files/usr/bin/env node`. 

When you run `node-red` in a normal session you are saved by the LD_PRELOAD'ed library [termux-exec](https://github.com/termux/termux-exec/), which automatically translates `/usr/bin -> /data/data/com.termux/files/usr/bin`. 

An alternative fix would be to not rely on the shebang at all in the termux-boot script, instead of having something like

```sh
#!/data/data/com.termux/files/usr/bin/sh
termux-wake-lock
node-red
```
You can do 
```
#!/data/data/com.termux/files/usr/bin/sh
termux-wake-lock
node ~/node_modules/.bin/node-red # change path if you installed globally
```

The prefixed `node` then overrides the shebang in the node-red script.

**Comment by Cjkeenan** [comment]
Sweet that seems to have fixed it. I just coped the path from `which node-red` and then put that path into the argument for the node command, see the whole script below.
```
#! /data/data/com.termux/files/usr/bin/sh
termux-wake-lock
node /data/data/com.termux/files/user/bin/node-red
```
Are there any issues with what I have done?

**Comment by Grimler91** [comment]
`/data/data/com.termux/files/user/bin/node-red` should be 
`/data/data/com.termux/files/usr/bin/node-red`

It should not have any other side-effects

**Comment by Cjkeenan** [comment]
yeah that was a typo on my part, it was actually usr in the script. Thank you for your help.

**Comment by tunerooster** [comment]
Not sure if this issue is still alive but...

I am running Android 10 on a Teclast M40.  I have followed all the instructions and advice from this thread and from the original documentation.  I won't bother with the details at this time...  but termux:boot is simply not starting.  I.e., all installed from f-droid, ran termux:boot (though there is no indication it does anything, other that show the instructions...  Is this expected?), script is in ~/.termux/boot, executable permissions, etc...

 Reboot... there is no output from logcat, or in the log file (created from my script with "set -x; exec > ~/boot.log 2>&1".  The script runs and creates the log when executed manually from termux shell.

Has anyone gotten this to work on Android 10?  If so, can you please advise...

Many thanks!

**Comment by MCorTHATMC** [comment]
I have the same problem with termux:boot doesn't have a session so that I can't ssh to termux. And I found that use "am start -n com.termux/.HomeActivity" can open termux so there would have a session (if you don't unlock your phone, the notification would still be 0 session but there seems to have a session because I can ssh to termux in this situation. ). You can add it into your start scripts in ~/.termux/boot/ .
As for running other commands, I use "screen -dm" and it worked. I founded that there exist the screens I created in ~/.termux/boot/start.sh.

> Has anyone gotten this to work on Android 10? If so, can you please advise...
I think this is a general problem happening to any Android version because I meet it on Android 7.1.

Pardon my bad English.

**Comment by Erbzur** [comment]
> I have the same problem with termux:boot doesn't have a session so that I can't ssh to termux. And I found that use "am start -n com.termux/.HomeActivity" can open termux so there would have a session (if you don't unlock your phone, the notification would still be 0 session but there seems to have a session because I can ssh to termux in this situation. ). You can add it into your start scripts in ~/.termux/boot/ . As for running other commands, I use "screen -dm" and it worked. I founded that there exist the screens I created in ~/.termux/boot/start.sh.
> 
> > Has anyone gotten this to work on Android 10? If so, can you please advise...
> 
> I think this is a general problem happening to any Android version because I meet it on Android 7.1.
> 
> Pardon my bad English.

The same not working on Android 7.1, after upgrading termux. My termux-boot is still on version 0.7. It did startup after system boot according to permission manager logs.

**Comment by Xavron** [comment]
> > Has anyone gotten this to work on Android 10? If so, can you please advise...
> 
> I think this is a general problem happening to any Android version because I meet it on Android 7.1.

I'm using Android 12 non-rooted and Termux boot works fine. You must do the scripts inside Termux and not device root of user storage which is easily lost understanding. The scripts will otherwise not be found. It should work on all devices. It just needs the script at the hard coded Termux app only path.

It could be that specific devices are blocking boot receiver or some other odd thing. You have to narrow that down. But if you do logcat verbose and filter to "termux" and wait, then termux boot should at least show up if the device is allowing the app to start on boot. If it starts on boot then it should have a message about no scripts or some other further info to go on.

**Comment by uwe-007** [comment]
I have similar issues on my Android 7 phone with Termux-boot-Addon.

First of all: 
If sshd is started via ~/.termux/boot/start.sh, it takes at least about 2 minutes more than starting sshd via ~/.bashrc (and Termux-App via "Launch on Boot" (https://f-droid.org/de/packages/news.androidtv.launchonboot/).
But I need to disable SIM PIN entering at device start  for this way (otherwise .bashrc will not be executed).

If starting with Termux-Boot-Addon this one works:
```
#!/data/data/com.termux/files/usr/bin/sh
termux-wake-lock
sshd
```
This one not:
```
#!/data/data/com.termux/files/usr/bin/sh
termux-wake-lock
. $PREFIX/etc/profile
sv start sshd ## ??? not sure if needed
```

**Comment by uwe-007** [comment]
One more curious (on Android 10):
This one works (sshd running):
`#!/data/data/com.termux/files/usr/bin/bash

termux-wake-lock

date > /data/data/com.termux/files/home/boot.log

sleep 2

sshd -p 2222`

After inserting the line "pkill sshd" sshd doesn't run after device restart.
`#!/data/data/com.termux/files/usr/bin/bash

termux-wake-lock

date > /data/data/com.termux/files/home/boot.log

pkill sshd

sleep 2

sshd -p 2222`

**Comment by Grimler91** [comment]
@uwe-007 if you have enabled sshd at some point (`sv-enable sshd`) then this should work:

```
$ cat .termux/boot/start-services
#!/data/data/com.termux/files/usr/bin/sh
termux-wake-lock
source /data/data/com.termux/files/usr/etc/profile.d/start-services.sh
```

(The wiki needs to be updated)

**Comment by uwe-007** [comment]
@Grimler91 
Thanks, Henrik, I will give it a try.
And why can't I use pkill (second post)?

**Comment by Grimler91** [comment]
@uwe-007 not sure, `adb logcat` from a computer when phone is connected and usb debugging enabled should show logs with error message

**Comment by islemchater** [comment]
> 

https://github.com/termux/termux-boot/issues/47#issue-665359642

**Comment by whatamesss** [comment]
so, its the weirdest thing....it worked perfectly yesterday and i rebooted numerous times to test it.
today, my script is not being executed. i tried various shebangs...yesterday it actually worked without any.
i'm calling an adb command. if i execute my shell script manually from termux, it works fine.

whats going on?

**Comment by trancephorm** [comment]
Here I have the case when it used to work, and it just stopped. I had to start Termux:Boot again to make it work. Running Samsung S22 with updated system, I'm here if anyone needs any logs.

**Comment by starb724600-ship-it** [comment]
$ cat 00_autostart 
#!/data/data/com.termux/files/usr/bin/bash
termux-wake-lock
date > $HOME/date
